DROOLS-1531 Fix instanceof accordingly to FEEL spec table 49 (optimizations)

tarilabs opened the pull request:
@etirelli please consider this PR for the 2 remaining comments which were noted here: https://github.com/kiegroup/drools/pull/1214#pullrequestreview-34157657

etirelli commented on kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/impl/MapBackedType.java:
```diff
@@ -47,19 +47,14 @@ public MapBackedType addField(String name, Type type) {
         for ( Entry<String, Type> f : fields.entrySet() ) {
-            if ( instance.get(f.getKey()) == null ) {
+            Object instanceValueForKey = instance.get(f.getKey());
+            if ( instance.get(f.getKey()) == null || !f.getValue().isInstanceOf(instanceValueForKey) ) {
```
@tarilabs is this correct? There are two cases, when a map backed instance does not have an a given attribute, and when it has the attribute but the value is null.
For instance, a Person instance might require the "Last Name" attribute to be considered a person, but the attribute might be set to null. It is still a person if the last name is null, but in duck typing it is not a person if the last name attribute does not exist.

etirelli commented on kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/impl/MapBackedType.java:
```diff
@@ -70,19 +65,14 @@ public boolean isAssignableValue(Object value) {
         for ( Entry<String, Type> f : fields.entrySet() ) {
-            if ( instance.get(f.getKey()) == null ) {
+            Object instanceValueForKey = instance.get(f.getKey());
+            if ( instanceValueForKey == null || !f.getValue().isAssignableValue(instanceValueForKey) ) {
```
same here

tarilabs commented:
@etirelli you are absolutely right, I'm just a little lost as I thought this is what I did originally with the two separate if checks, possibly I lost it even before during some refactoring.. checking.

tarilabs commented:
@etirelli I've updated following your comments, and the new test manifested an issue which was present on master even before (https://github.com/kiegroup/drools/blob/master/kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/impl/MapBackedType.java#L30) , now corrected.

etirelli commented:
Looks good @tarilabs ! Thanks!
